### PR TITLE
cdc: handles region registers correctly after connection deregistered

### LINKED ISSUE #16444: cdc endpoint thread panics
### What did happened?
```
[2024/01/17 23:04:04.645 +00:00] [WARN] [service.rs:281] ["cdc receive failed"] [conn_id=ConnId(122980)] [downstream=ipv4:172.20.81.227:53050] [error=RpcFinished(None)]
[2024/01/17 23:04:04.647 +00:00] [WARN] [service.rs:281] ["cdc receive failed"] [conn_id=ConnId(122982)] [downstream=ipv4:172.20.81.227:53052] [error=RpcFinished(None)]
[2024/01/17 23:04:04.648 +00:00] [WARN] [service.rs:281] ["cdc receive failed"] [conn_id=ConnId(122981)] [downstream=ipv4:172.20.81.227:53050] [error=RpcFinished(None)]
[2024/01/17 23:04:04.649 +00:00] [INFO] [service.rs:278] ["cdc receive closed"] [conn_id=ConnId(122983)] [downstream=ipv4:172.20.81.227:53052]
[2024/01/17 23:04:04.657 +00:00] [INFO] [service.rs:278] ["cdc receive closed"] [conn_id=ConnId(122984)] [downstream=ipv4:172.20.81.227:53050]
[2024/01/17 23:04:04.658 +00:00] [FATAL] [lib.rs:509] ["called `Option::unwrap()` on a `None` value"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:508:18\n   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2032:9\n      std::panicking::rust_panic_with_hook\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:692:13\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:577:13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:137:18\n   4: rust_begin_unwind\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:575:5\n   5: core::panicking::panic_fmt\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:65:14\n   6: core::panicking::panic\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:114:5\n   7: cdc::endpoint::Endpoint<T,E>::on_register\n      <cdc::endpoint::Endpoint<T,E> as tikv_util::worker::pool::Runnable>::run\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/cdc/src/endpoint.rs:1143:18\n   8: tikv_util::worker::pool::Worker::start_with_timer_impl::{{closure}}\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/worker/pool.rs:495:25\n      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:91:19\n      yatp::task::future::RawTask<F>::poll\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/task/future.rs:59:9\n   9: yatp::task::future::TaskCell::poll\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/task/future.rs:103:9\n      <yatp::task::future::Runner as yatp::pool::runner::Runner>::handle\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/task/future.rs:387:20\n  10: <tikv_util::yatp_pool::YatpPoolRunner<T> as yatp::pool::runner::Runner>::handle\n             at home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/yatp_pool/mod.rs:122:24\n      yatp::pool::worker::WorkerThread<T,R>::run\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/pool/worker.rs:48:13\n      yatp::pool::builder::LazyBuilder<T>::build::{{closure}}\n             at rust/git/checkouts/yatp-e704b73c3ee279b6/39cb495/src/pool/builder.rs:114:25\n      std::sys_common::backtrace::__rust_begin_short_backtrace\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:121:18\n  11: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:551:17\n      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:271:9\n      std::panicking::try::do_call\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:483:40\n      std::panicking::try\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:447:19\n      std::panic::catch_unwind\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:137:14\n      std::thread::Builder::spawn_unchecked_::{{closure}}\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:550:30\n      core::ops::function::FnOnce::call_once{{vtable.shim}}\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:513:5\n  12: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2000:9\n      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2000:9\n      std::sys::unix::thread::Thread::new::thread_start\n             at rust/toolchains/nightly-2022-11-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17\n  13: <unknown>\n  14: clone\n"] [location=/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/cdc/src/endpoint.rs:635] [thread_name=cdc-0]

```

### PR DESCRIPTION
**Pull request by hicqu**
### What is changed and how it works?

Issue Number: Close #16444 

What's Changed:

```commit-message
cdc: handles region registers correctly after connection deregistered
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/endpoint.rs [hunk 1]
```diff
@@ -697,8 +697,13 @@ impl<T: 'static + CdcHandle<E>, E: KvEngine, S: StoreRegionMeta> Endpoint<T, E,
...
         let downstream_state = downstream.get_state();
 
-        // Register must follow OpenConn, so the connection must be available.
-        let conn = self.connections.get_mut(&conn_id).unwrap();
+        // The connection can be deregistered by some internal errors. Clients will
+        // be always notified by closing the GRPC server stream, so it's OK to drop
+        // the task directly.
+        let conn = match self.connections.get_mut(&conn_id) {
+            Some(conn) => conn,
+            None => return,
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs:705]: Shall we add some log here?

**Review by overvenus — APPROVED** [review]

**Review by asddongmen — APPROVED** [review]

**Review by sdojjy — APPROVED** [review]

**Review by YuJuncen — APPROVED** [review]

**Review by 3AceShowHand — APPROVED** [review]

**Pull request merged**